# skin_pQTL Analysis

This repository contains the code and essential data used for the analysis in the skin pQTL study.  
The project includes R scripts, shell workflows, and  datasets.  

---

##  Overview

The scripts in this repository cover the complete analysis workflow:

1. **Proteome processing and QC**
2. **Age-associated proteome analysis**
3. **pQTL discovery and fine-mapping**
4. **QTL post-processing and annotation**
5. **Transcriptome-wide and proteome-wide association analyses (TWAS/PWAS)**
6. **Integration with disease GWAS (e.g., leprosy)**


---

## Data Availability

The `data/` folder contains **essentail files** to illustrate input formats required to reproduce all results  .  
** Some full datasets are not completely included** due to size constraints (>50â€“100 MB).


To fully reproduce the results, please download or generate the required data.

Links for downloading full data or instructions to generate them may be provided in R or shell scripts.

---

## Usage

### 1. Running the R scripts
Navigate to the `R/` folder.  
Scripts are named following the analysis workflow (e.g., preprocessing â†’ pQTL mapping â†’ integration â†’ visualization).

Example:
```bash
Rscript 01_proteome_processing.R
Rscript 04_QTL_process.R

2. Running the shell scripts

Shell scripts in shell/ folder are used for:

batch processing

submitting jobs to HPC clusters

running QTL tools

Example:

bash 03_pQTL_formal_code.sh

Ensure required software and environment modules are installed.


---

## âš™ï¸ Dependencies

- **R** (â‰¥ 4.2.0) with packages:  
  `tidyverse`, `data.table`, `ggplot2`, `MatrixEQTL`, `plink2R`, `coloc`, `qvalue`
- **Python** (optional, for specific utilities)
- **QTLtools** or **TensorQTL** for pQTL mapping
- **FUSION** for TWAS
- **PLINK 1.9/2.0**
- **LocusZoom** or **ggplot2** for visualization

---

Modify input/output paths in each script as needed.

ğŸ“„ License

This project is released under the MIT License.
You are free to reuse and adapt the code with appropriate citation.

ğŸ“š Citation
(Manuscript in preparation.)

ğŸ“¬ Contact

For questions or collaborations, please contact:
[Yuanqiang Sun] (sunyuanqiang[AT]pku.edu.cn)
